File: src/python/Rumble/rumble.py
#!/usr/bin/env python3
"""
Author:      Ceyhun Uzunoglu <ceyhunuzngl AT gmail dot com>
Description: Client to run jsoniq queries in rumble server.
"""

# system modules
import os
import json
import argparse

# third party libs
import requests


class OptionParser:
    def __init__(self):
        """ User based option parser """
        # Max output size is unlimited
        default_url = "https://cms-monitoring.cern.ch/"
        uri = default_url + "jsoniq?overwrite=yes&materialization-cap=-1"
        self.parser = argparse.ArgumentParser(prog='PROG')
        self.parser.add_argument("--query", action="store", dest="query", default="", help="Input: file or query")
        self.parser.add_argument("--output", action="store", dest="output", default=None, help="Output file")
        self.parser.add_argument("--server", action="store", dest="server", default=uri, help="Rumble server")


def handle_output_file(output_file):
    """ Handles output file """
    # Accepts both given full path of file or
    # only file name which will be created in current directory.
    if os.path.isdir(os.path.abspath(os.path.join(output_file, os.pardir))):
        return output_file
    else:
        print("Given output file path is not exist or not reachable:", output_file)
        print("Writing to default path: ", os.path.join(os.getcwd(), "rumble_output.txt"))
        return os.path.join(os.getcwd(), "rumble_output.txt")


def rumble(server, query, output=None):
    response = json.loads(requests.post(server, data=query).text)
    if 'warning' in response:
        print(json.dumps(response['warning']))
    if 'values' in response:
        if output:
            output_file = handle_output_file(output)
            try:
                with open(output_file, "w+") as f:
                    f.write(json.dumps(response))
                    return "Successfully written to: " + str(os.path.join(os.getcwd(), output))
            except Exception as e:
                print("Could not write to file:", output_file)
                print(e)
        else:
            for e in response['values']:
                return json.dumps(e)
    elif 'error-message' in response:
        return response['error-message']
    else:
        return response


def main():
    """ Main function """
    optmgr = OptionParser()
    opts = optmgr.parser.parse_args()
    if os.path.exists(opts.query):  # Get query from file
        server = opts.server
        query = open(opts.query).read()
        print("<Input Query>:\n" + str(query))
    elif opts.query.startswith("hdfs://"):
        server = opts.server + "&query-path=" + opts.query
        query = None
        print("<Input Query in HDFS>:", opts.query)
    else:
        server = opts.server
        query = opts.query
        print("<Input Query>:\n" + str(query))
    # Get output path as output
    if opts.output:
        output = opts.output
    else:
        print("Output path is not given. Result will be stdout!")
        output = None
    resp = rumble(server, query, output)
    if resp:
        print("<Rumble Response Values>")
        print(resp)


if __name__ == '__main__':
    main()

File: src/go/Rumble/rumble.go
package main

// File       : rumble_go.go
// Author     : Ceyhun Uzunoglu <cuzunogl AT gmail dot com>
// Description: client for rumble server

import (
	"flag"
	"fmt"
	"io"
	"io/ioutil"
	"log"
	"net/http"
	"net/http/httputil"
	"os"
	"strings"
)

// helper function to either read file content or return given string. Second parameter is for isHdfs.
func read(r string) (string, bool) {
	// if query is in hdfs, use query-path parameter in Rumble
	// if not, read query from file or give query string itself
	if strings.HasPrefix(r, "hdfs://") {
		return r, true
	} else if _, err := os.Stat(r); err == nil {
		b, e := ioutil.ReadFile(r)
		if e != nil {
			log.Fatalf("Unable to read data from file: %s, error: %s", r, e)
		}
		return strings.Replace(string(b), "\n", "", -1), false
	}
	return r, false
}

func write(resp *http.Response, output string) {
	// if output is not given, print response to stdout
	if output == "" {
		data, err := ioutil.ReadAll(resp.Body)
		if err != nil {
			log.Fatal(err)
		}
		log.Println("RESPONSE: \n" + string(data))
	} else {
		outFile, err := os.Create(output)
		if err != nil {
			log.Fatal("ERROR: unable to create file:", err)
		}
		defer outFile.Close()
		io.Copy(outFile, resp.Body)
		log.Println("Response is written to file:", output)
	}
}

// helper function to send Rumble request
func sendRequest(server string, query string, output string, isHdfs bool, verbose int) {
	var req *http.Request
	var err error
	rurl := server
	if verbose > 0 {
		log.Println("Server:", rurl)
		log.Println("Query:\n" + query)
	}

	if !isHdfs {
		// if JSONiq query is not in hdfs file
		req, err = http.NewRequest("POST", rurl, strings.NewReader(query))
	} else {
		rurl := fmt.Sprintf("%s&overwrite=yes&query-path=%s", server, query)
		req, err = http.NewRequest("POST", rurl, nil)
	}

	if err != nil {
		log.Fatalf("Unable to make request to %s, error: %s", rurl, err)
	}

	err = nil // delete
	if verbose > 1 {
		dump, err := httputil.DumpRequestOut(req, true)
		if err == nil {
			log.Println("Request:\n" + string(dump))
		}
	}

	client := &http.Client{}
	resp, err := client.Do(req)
	if err != nil {
		log.Fatalf("Unable to get response from %s, error: %s", rurl, err)
	}
	defer resp.Body.Close()

	log.Println("Response Status:", resp.Status)
	if verbose > 0 {
		log.Println("Response Headers:", resp.Header)
	}
	if resp.StatusCode == http.StatusOK {
		write(resp, output)
	}
}

func main() {
	defaultUrl := "https://cms-monitoring.cern.ch/"
	uri := defaultUrl + "jsoniq?materialization-cap=-1"
	var verbose int
	flag.IntVar(&verbose, "verbose", 0, "verbosity level")
	var server string
	flag.StringVar(&server, "server", uri, "Rumble server URL")
	var query string
	flag.StringVar(&query, "query", "", "query string or query JSONiq file")
	var output string
	flag.StringVar(&output, "output", "", "specify output file path, if not defined, print to stdout")
	flag.Usage = func() {
		fmt.Println("Usage: rumble_go [options]")
		flag.PrintDefaults()
		fmt.Println("Examples:")
		fmt.Println("   # simple rumble query, prints to stdout")
		fmt.Println("   rumble_go -query=\"1+1\"")
		fmt.Println("")
		fmt.Println("   # write result to file")
		fmt.Println("   rumble_go -query=\"1+1\" -output output.txt")
		fmt.Println("")
		fmt.Println("   # get query from hdfs or file and write result to file")
		fmt.Println("   rumble_go -query=\"./exmaple.jq\" -output output.txt")
		fmt.Println("   rumble_go -query=\"hdfs://anlaytix/rumble.jq\" -output output.txt")
		fmt.Println("")
	}
	flag.Parse()
	if verbose > 0 {
		log.SetFlags(log.LstdFlags | log.Lshortfile)
	} else {
		log.SetFlags(log.LstdFlags)
	}
	q, isHdfs := read(query)
	sendRequest(server, q, output, isHdfs, verbose)
}
